src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -817,59 +839,110 @@ public void ReceivingMessage()
 
                 _stopwatch = Stopwatch.StartNew();
             }
+
+            private ResponseMessage ReadResponseMessage(ByteBufferStream stream, IMessageEncoderSelector encoderSelector)
+            {
+                var encoderFactory = new BinaryMessageEncoderFactory(stream, _messageEncoderSettings, _compressorSource);
+                var selector = new CompressedMessageEncoderSelector(encoderSelector);
+                var message = selector.GetEncoder(encoderFactory).ReadMessage();
+
+                return message is CompressedMessage compressedMessage
+                    ? (ResponseMessage)compressedMessage.OriginalMessage
+                    : (ResponseMessage)message;
+            }
         }
 
         private class SendMessagesHelper
         {
+            private readonly ICompressorSource _compressorSource;
```

Comment:
Undo all the changes to SendMessageHelper.

Let's leave this class the same as before. It is only responsible for encoding the original message, not for compression.